File: content/6-rag/1-setup/mastra/1-setup.ts
import { LocalIndex } from "vectra";
import { readFileSync } from "fs";
import OpenAI from "openai";

// Vectra is a local vector database that stores documents with their embeddings.
// Unlike ChromaDB (Python), Vectra requires you to compute embeddings yourself —
// we use OpenAI's embedding model for that.
const openai = new OpenAI();

// LocalIndex persists to disk so agent scripts can query it later
const index = new LocalIndex("./acme_index");
if (!(await index.isIndexCreated())) await index.createIndex();

// load documents from docs.txt — each non-empty line becomes one document.
// in production you'd load real files, DB rows, or API responses instead.
const lines = readFileSync("docs.txt", "utf-8")
  .split("\n")
  .filter((l) => l.trim() && !l.startsWith("#"));

// embed each document and store it in the index
for (const [i, doc] of lines.entries()) {
  const res = await openai.embeddings.create({
    model: "text-embedding-3-small",
    input: doc,
  });
  await index.insertItem({
    id: `doc_${i}`,
    metadata: { text: doc },
    vector: res.data[0].embedding,
  });
}
console.log(`Indexed ${lines.length} documents`);

File: content/6-rag/1-setup/anthropic/1-setup.py
from pathlib import Path
import os
import chromadb
import chromadb.utils.embedding_functions as embedding_functions

# ChromaDB is a vector database that stores documents and their embeddings.
# PersistentClient saves the index to disk so agent scripts can query it later.
client = chromadb.PersistentClient(path="./acme_index")

# tell ChromaDB which embedding model to use for indexing and queries.
# without this, it silently downloads an ~80MB local model on first run.
openai_ef = embedding_functions.OpenAIEmbeddingFunction(
    api_key=os.environ["OPENAI_API_KEY"],
    model_name="text-embedding-3-small",
)

# a collection groups related documents — the embedding function above
# converts text to vectors automatically during add and query operations
collection = client.get_or_create_collection(
    "acme_docs", embedding_function=openai_ef,
)

# load documents from docs.txt — each non-empty line becomes one document.
# in production you'd load real files, DB rows, or API responses instead.
lines = [l for l in Path("docs.txt").read_text().splitlines()
         if l.strip() and not l.startswith("#")]

# upsert inserts new docs or updates existing ones (safe to re-run)
collection.upsert(
    documents=lines,
    ids=[f"doc_{i}" for i in range(len(lines))],
)
print(f"Indexed {len(lines)} documents")
